### PR TITLE
Basic scaffolding for OpenCL

### PR DESCRIPTION
**Pull request by bwasti**
test it out:
mkdir build && cd build && cmake -DUSE_OPENCL=1 .. && make -j24 && cd .. && PYTHONPATH=build python caffe2/contrib/opencl/test/conv_test.py

### CODE REVIEW

**Comment by bwasti** [comment]
this includes some convolution implementations that are reasonably efficient on mobile

**Comment by psyhtest** [comment]
@bwasti Out of curiosity, which mobile platforms have you tested your implementation on?

**Comment by bwasti** [comment]
@psyhtest Galaxy S7 (international and US versions).  However, before pushing to my branch I had to change some code for compat with a Mac Pro, so I haven't tested the latest on these devices at all.

**Comment by psyhtest** [comment]
@bwasti, good choice! Adreno (S7 US) and Mali (S7 International) are the two dominant mobile GPU architectures. However, the Mali Midgard architecture (e.g. Mali-T880) has been superseded by the Mali Bifrost architecture (e.g. Mali-G71), so I would advise to test also on e.g. S8 International.

**Comment by bwasti** [comment]
@psyhtest that's good to know, thanks!

File: .gitmodules [hunk 1]
```diff
@@ -49,3 +49,6 @@
 [submodule "third_party/NNPACK_deps/psimd"]
 	path = third_party/NNPACK_deps/psimd
 	url = https://github.com/Maratyszcza/psimd.git
+[submodule "third_party/libopencl-stub"]
+	path = third_party/libopencl-stub
+	url = git@github.com:bwasti/libopencl-stub.git
```
Comment (Yangqing) [inline .gitmodules:54]: Should we do git@ or https?

File: caffe2/contrib/opencl/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# ---[ OpenCL
+if(USE_OPENCL)
+  set(Caffe2_CPU_SRCS ${Caffe2_CPU_SRCS} ${CMAKE_SOURCE_DIR}/third_party/libopencl-stub/src/libopencl.c)
```
Comment (Yangqing) [inline caffe2/contrib/opencl/CMakeLists.txt:3]: Would this cause duplicates symbols if some other libraries also compile it?

**Comment by orionr** [comment]
Thank you for your contribution! PyTorch and Caffe2 are now officially merged with the PyTorch 1.0 preview release. Please submit on https://github.com/pytorch/pytorch if this is still an issue over there.

**Pull request closed without merging**